platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+It details the roster's lifecycle, data storage, API, DevOps workflow, and changes necessary changes to associated
+components within the platform.
+
+| Metadata           | Entities                                      | 
+|--------------------|-----------------------------------------------|
+| Designers          | Kore, Anthony, Ed, Austin                     |
+| Functional Impacts | TSS, Address Book, Reconnect,                 |
+| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
+| HIPS               | N/A                                           |
+
+## Purpose and Context
+
+The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
+consensus and block signing.
+The Roster, an immutable subset of the address book, will provide the data structure and API for this mechanism.
+This proposal provides a specification for the behavior of rosters starting from their creation from the Address Book to
+their terminal state within the TSS specification.
+A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.
+The candidate roster will always get adopted on the next software upgrade.
+
+The Hedera App (henceforth reffered to as 'App') will maintain a version of the Address Book as a dynamic list that
+reflects the desired future state of
+the
+network's nodes. This dynamic list of Addresses will be continuously updated by HAPI transactions, such as those that
+create, update, or delete nodes.
+At some point when the App decides it's time to begin work on adopting a new address book (scope beyond this proposal),
+it will create a `Roster` object with information from the Address Book and pass it to the platform.
+
+The mechanism for doing so is detailed below.
+
+## What this proposal does not address
+
+## Requirements
+
+- Immutability: The roster hash will be a critical piece of data. Rosters must be immutable to protect the integrity
+  of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the App to create and submit candidate rosters to the platform,
+  and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use the new `Roster` object instead.
+- Address book paradigm discontinued within the platform codebase in favor of `Roster`s
+
+## Architecture
+
+###### Roster Lifecycle
+
+![](TSS%20Roster%20Lifecycle.drawio.svg)
+
+### Roster Public API
+
+#### Setting the Candidate Roster
+A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+
+```java
+//in PlaformState
```

Comment:
This is no longer accurate. The roster will be added to the RosterMap and the hash of the candidate roster will be stored in the RosterState singleton https://app.diagrams.net/#G1OD12O2KqycNWIzKu6vP-ExPEGKH9-8De#%7B%22pageId%22%3A%22KP1hL-13WYacigEbZWgY%22%7D

Reply:
Current approach is to have Services call this method with a Roster. Once done, subsequent routines are called to validate the Roster, compute the hash and persist both the Roster and hash in the various data structures in the durable state. This is still valid. We need to be able to reject the Roster synchronously if invalid.